File: app/src/main/java/com/mandarin/bcu/androidutil/io/ErrorLogWriter.java
package com.mandarin.bcu.androidutil.io;

import android.os.Build;
import android.os.Environment;

import com.mandarin.bcu.androidutil.StaticStore;

import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class ErrorLogWriter implements Thread.UncaughtExceptionHandler {
    private Thread.UncaughtExceptionHandler errors;
    private String path;
    private boolean upload;

    public ErrorLogWriter(String path, boolean upload) {
        this.path = path;
        this.errors = Thread.getDefaultUncaughtExceptionHandler();
        this.upload = upload;
    }

    @Override
    public void uncaughtException(Thread t, Throwable e) {
        if (path != null) {
            writeToFile(e);
        }

        errors.uncaughtException(t, e);
    }

    private void writeToFile(Throwable e) {
        try {
            File f = new File(path);

            if (!f.exists()) {
                f.mkdirs();
            }

            File fe = new File(Environment.getDataDirectory()+"/data/com.mandarin.bcu/upload");

            if(!fe.exists()) fe.mkdirs();

            SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy_MM_dd_HH_mm_ss", Locale.US);
            Date date = new Date();
            String name = dateFormat.format(date);

            final Writer stringbuff = new StringWriter();
            final PrintWriter printWriter = new PrintWriter(stringbuff);
            e.printStackTrace(printWriter);
            String current = stringbuff.toString();
            printWriter.close();

            if(upload) {
                String dname = name+"_"+Build.MODEL+".txt";

                File df = new File(Environment.getDataDirectory() + "/data/com.mandarin.bcu/upload/", dname);

                FileWriter dfileWriter = new FileWriter(df);
                dfileWriter.append("VERSION : ").append(StaticStore.VER).append("\r\n");
                dfileWriter.append("MODEL : ").append(Build.MANUFACTURER).append(" ").append(String.valueOf(Build.MODEL)).append("\r\n");
                dfileWriter.append("IS EMULATOR : ").append(String.valueOf(Build.MODEL.contains("Emulator") || Build.MODEL.contains("Android SDK"))).append("\r\n");
                dfileWriter.append("ANDROID_VER : ").append("API ").append(String.valueOf(Build.VERSION.SDK_INT)).append(" (").append(Build.VERSION.RELEASE).append(")").append("\r\n").append("\r\n");
                dfileWriter.append(current);
                dfileWriter.flush();
                dfileWriter.close();
            }

            name += ".txt";

            File file = new File(path, name);

            if (!file.exists())
                f.createNewFile();

            FileWriter fileWriter = new FileWriter(file);
            fileWriter.append(current);
            fileWriter.flush();
            fileWriter.close();

        } catch (IOException e1) {
            e1.printStackTrace();
        }
    }

    public static void WriteLog(Exception error, boolean upload) {
        try {
            String path = Environment.getExternalStorageDirectory().getPath() + "/BCU/logs/";
            File f = new File(path);

            if (!f.exists()) {
                f.mkdirs();
            }

            SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy_MM_dd_HH_mm_ss", Locale.US);
            Date date = new Date();
            String name = dateFormat.format(date) + ".txt";

            File file = new File(path, name);

            final Writer stringbuff = new StringWriter();
            final PrintWriter printWriter = new PrintWriter(stringbuff);

            error.printStackTrace(printWriter);

            if (!file.exists())
                file.createNewFile();
            else {
                file = new File(path, GetExistingFileName(path, name));
                file.createNewFile();
            }

            if(upload) {
                File df = new File(Environment.getDataDirectory() + "/data/com.mandarin.bcu/upload/", name);

                if(!df.exists()) df.createNewFile();

                FileWriter dfileWriter = new FileWriter(df);
                dfileWriter.append("VERSION : ").append(StaticStore.VER).append("\r\n");
                dfileWriter.append("MODEL : ").append(Build.MANUFACTURER).append(" ").append(String.valueOf(Build.MODEL)).append("\r\n");
                dfileWriter.append("IS EMULATOR : ").append(String.valueOf(Build.MODEL.contains("Emulator") || Build.MODEL.contains("Android SDK"))).append("\r\n");
                dfileWriter.append("ANDROID_VER : ").append("API ").append(String.valueOf(Build.VERSION.SDK_INT)).append(" (").append(Build.VERSION.RELEASE).append(")").append("\r\n").append("\r\n");
                dfileWriter.append(stringbuff.toString());
                dfileWriter.flush();
                dfileWriter.close();
            }

            FileWriter fileWriter = new FileWriter(file);
            fileWriter.append(stringbuff.toString());
            fileWriter.flush();
            fileWriter.close();
            printWriter.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static String GetExistingFileName(String path, String name) {
        boolean decided = false;

        int exist = 1;

        String nam = name + "-" + exist;

        while (!decided) {
            File f = new File(path, nam);

            if (!f.exists())
                return nam;
            else {
                exist++;
                nam = name + "-" + exist;
            }

            decided = true;
        }

        return nam;
    }

    public static boolean Upload(File file) throws IOException {
        if (!file.exists()) return false;

        String crlf = "\r\n";
        String hyphens = "--";
        String bound = "*****";

        FileInputStream fileInputStream = new FileInputStream(file);
        URL u = new URL("https://battle-cats-ultimate.000webhostapp.com/api/java/alogio.php");

        HttpURLConnection con = (HttpURLConnection) u.openConnection();
        con.setDoInput(true);
        con.setDoOutput(true);
        con.setUseCaches(false);
        con.setRequestMethod("POST");
        con.setRequestProperty("Connection", "Keep-Alive");
        con.setRequestProperty("ENCTYPE", "multipart/form-data");
        con.setRequestProperty("Content-Type", "multipart/form-data;boundary=*****");
        con.setRequestProperty("uploaded_file", file.getName());

        DataOutputStream dos = new DataOutputStream(con.getOutputStream());

        byte[] buffer;
        int available, read, size;
        int max = 1024*1024;

        dos.writeBytes(hyphens+bound+crlf);
        dos.writeBytes("Content-Disposition: form-data; name=\""+"catFile"+"\";filename=\""+file.getName()+"\""+crlf);
        dos.writeBytes(crlf);

        available = fileInputStream.available();
        size = Math.min(available, max);
        buffer = new byte[size];

        read = fileInputStream.read(buffer, 0, size);

        while (read > 0) {
            dos.write(buffer, 0, size);
            available = fileInputStream.available();
            size = Math.min(available, size);
            read = fileInputStream.read(buffer, 0, size);
        }

        dos.writeBytes(crlf);
        dos.writeBytes(hyphens+bound+crlf);

        int response = con.getResponseCode();

        dos.flush();
        dos.close();
        fileInputStream.close();
        con.disconnect();

        return response == 200;
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/io/ErrorLogWriter.kt
package com.mandarin.bcu.androidutil.io

import android.os.Build
import android.os.Environment
import com.mandarin.bcu.androidutil.StaticStore
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

class ErrorLogWriter(private val path: String?, private val upload: Boolean) : Thread.UncaughtExceptionHandler {
    private val errors: Thread.UncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()
    override fun uncaughtException(t: Thread, e: Throwable) {
        if (path != null) {
            writeToFile(e)
        }
        errors.uncaughtException(t, e)
    }

    private fun writeToFile(e: Throwable) {
        try {
            val f = File(path)
            if (!f.exists()) {
                f.mkdirs()
            }
            val fe = File(Environment.getDataDirectory().toString() + "/data/com.mandarin.bcu/upload")
            if (!fe.exists()) fe.mkdirs()
            val dateFormat = SimpleDateFormat("yyyy_MM_dd_HH_mm_ss", Locale.US)
            val date = Date()
            var name = dateFormat.format(date)
            val stringbuff: Writer = StringWriter()
            val printWriter = PrintWriter(stringbuff)
            e.printStackTrace(printWriter)
            val current = stringbuff.toString()
            printWriter.close()
            if (upload) {
                val dname = name + "_" + Build.MODEL + ".txt"
                val df = File(Environment.getDataDirectory().toString() + "/data/com.mandarin.bcu/upload/", dname)
                val dfileWriter = FileWriter(df)
                dfileWriter.append("VERSION : ").append(StaticStore.VER).append("\r\n")
                dfileWriter.append("MODEL : ").append(Build.MANUFACTURER).append(" ").append(Build.MODEL.toString()).append("\r\n")
                dfileWriter.append("IS EMULATOR : ").append((Build.MODEL.contains("Emulator") || Build.MODEL.contains("Android SDK")).toString()).append("\r\n")
                dfileWriter.append("ANDROID_VER : ").append("API ").append(Build.VERSION.SDK_INT.toString()).append(" (").append(Build.VERSION.RELEASE).append(")").append("\r\n").append("\r\n")
                dfileWriter.append(current)
                dfileWriter.flush()
                dfileWriter.close()
            }
            name += ".txt"
            val file = File(path, name)
            if (!file.exists()) f.createNewFile()
            val fileWriter = FileWriter(file)
            fileWriter.append(current)
            fileWriter.flush()
            fileWriter.close()
        } catch (e1: IOException) {
            e1.printStackTrace()
        }
    }

    companion object {
        fun writeLog(error: Exception, upload: Boolean) {
            try {
                val path = Environment.getExternalStorageDirectory().path + "/BCU/logs/"
                val f = File(path)
                if (!f.exists()) {
                    f.mkdirs()
                }
                val dateFormat = SimpleDateFormat("yyyy_MM_dd_HH_mm_ss", Locale.US)
                val date = Date()
                val name = dateFormat.format(date) + ".txt"
                var file = File(path, name)
                val stringbuff: Writer = StringWriter()
                val printWriter = PrintWriter(stringbuff)
                error.printStackTrace(printWriter)
                if (!file.exists()) file.createNewFile() else {
                    file = File(path, getExistingFileName(path, name))
                    file.createNewFile()
                }
                if (upload) {
                    val df = File(Environment.getDataDirectory().toString() + "/data/com.mandarin.bcu/upload/", name)
                    if (!df.exists()) df.createNewFile()
                    val dfileWriter = FileWriter(df)
                    dfileWriter.append("VERSION : ").append(StaticStore.VER).append("\r\n")
                    dfileWriter.append("MODEL : ").append(Build.MANUFACTURER).append(" ").append(Build.MODEL.toString()).append("\r\n")
                    dfileWriter.append("IS EMULATOR : ").append((Build.MODEL.contains("Emulator") || Build.MODEL.contains("Android SDK")).toString()).append("\r\n")
                    dfileWriter.append("ANDROID_VER : ").append("API ").append(Build.VERSION.SDK_INT.toString()).append(" (").append(Build.VERSION.RELEASE).append(")").append("\r\n").append("\r\n")
                    dfileWriter.append(stringbuff.toString())
                    dfileWriter.flush()
                    dfileWriter.close()
                }
                val fileWriter = FileWriter(file)
                fileWriter.append(stringbuff.toString())
                fileWriter.flush()
                fileWriter.close()
                printWriter.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        private fun getExistingFileName(path: String, name: String): String {
            var decided = false
            var exist = 1
            var nam = "$name-$exist"
            while (!decided) {
                val f = File(path, nam)
                nam = if (!f.exists()) return nam else {
                    exist++
                    "$name-$exist"
                }
                decided = true
            }
            return nam
        }

        @Throws(IOException::class)
        fun upload(file: File): Boolean {
            if (!file.exists()) return false
            val crlf = "\r\n"
            val hyphens = "--"
            val bound = "*****"
            val fileInputStream = FileInputStream(file)
            val u = URL("https://battle-cats-ultimate.000webhostapp.com/api/java/alogio.php")
            val con = u.openConnection() as HttpURLConnection
            con.doInput = true
            con.doOutput = true
            con.useCaches = false
            con.requestMethod = "POST"
            con.setRequestProperty("Connection", "Keep-Alive")
            con.setRequestProperty("ENCTYPE", "multipart/form-data")
            con.setRequestProperty("Content-Type", "multipart/form-data;boundary=*****")
            con.setRequestProperty("uploaded_file", file.name)
            val dos = DataOutputStream(con.outputStream)
            val buffer: ByteArray
            var available: Int
            var read: Int
            var size: Int
            val max = 1024 * 1024
            dos.writeBytes(hyphens + bound + crlf)
            dos.writeBytes("Content-Disposition: form-data; name=\"" + "catFile" + "\";filename=\"" + file.name + "\"" + crlf)
            dos.writeBytes(crlf)
            available = fileInputStream.available()
            size = available.coerceAtMost(max)
            buffer = ByteArray(size)
            read = fileInputStream.read(buffer, 0, size)
            while (read > 0) {
                dos.write(buffer, 0, size)
                available = fileInputStream.available()
                size = available.coerceAtMost(size)
                read = fileInputStream.read(buffer, 0, size)
            }
            dos.writeBytes(crlf)
            dos.writeBytes(hyphens + bound + crlf)
            val response = con.responseCode
            dos.flush()
            dos.close()
            fileInputStream.close()
            con.disconnect()
            return response == 200
        }
    }

}